src/rtpghi~.c

```diff
@@ -1,158 +1,243 @@
+	rtpghi_state* sta_pd;
+	ltfat_complex *c;
+	ltfat_int blocksize;
+	LTFAT_FIRWIN window;
     t_float f;
-    
-    t_sample **out;
-    
 } t_rtpghi_tilde;
 
-t_int *rtpghi_tilde_perform(t_int *w)
-{
-    t_rtpghi_tilde *x = (t_rtpghi_tilde *)(w[1]);
-    const float *s =      (const float*)(w[2]);
-    int            n =             (int)(w[3]);
-    
-    ltfat_complex_s *c = (ltfat_complex_s *) x->c;
-    
-    int e = 0;
-    
-    if ((s==NULL) || ((x->sta_pd)==NULL)) {
-        post("arrays not initialised");
-    }
-    else
-    {
-        e = phaseret_rtpghi_execute_s(x->sta_pd, s, c);
-        t_sample *out=(t_sample *) x->out[0];
-        t_sample *out1= (t_sample *) x->out[1];
-        //post("e: %d", e);
-        //post("complex signal: %f + i%f\n",creal(*(c+20)), cimag(*(c+20)));
-        while (n--) {
-            
-            *out = creal(*c);
-            *out1 = cimag(*c++);
-        }
-    }
-
-  return (w+4);
+t_int *rtpghi_tilde_perform(t_int *w) {
+	
+	t_rtpghi_tilde *x = (t_rtpghi_tilde *)(w[1]);
+	t_sample *s =      (t_sample *) (w[2]);
+	t_sample *out=(t_sample *) (w[3]);
+	t_sample *out1= (t_sample *) (w[4]);
+	int            n =             (int)(w[5]);
+
+	ltfat_complex *c = (ltfat_complex *) x->c;
+	int e = 0;
+	if (x->sta_pd == NULL || s==NULL) {
+		*out++ = 0.0;
+		*out1++ = 0.0;
+	}
+	else {
+		e = rtpghi_execute(x->sta_pd, s, c);
+	}
+
+	if (e == 0) {
+		while (n--) {
+			*out++ = creal(*c);
+			*out1++ = cimag(*c++);
+		}
+	}
+	else {
+		pd_error(x, "error of type: %d", e);
+	}
+	
+ return (w+6);
+	
 }
 
-void rtpghi_tilde_dsp(t_rtpghi_tilde *x, t_signal **sp)
-{
-    ltfat_int M = x->M_pd;
-    
-    //ltfat_complex_s *c = x->c;
-    post("length of c[]: %d", (M) * sizeof *(x->c));
-    post("s_n: %d", sp[0]->s_n);
-    
-    t_sample **dummy=x->out;
-    *dummy++=sp[1]->s_vec;
-    *dummy=sp[2]->s_vec;
-    
-    dsp_add(rtpghi_tilde_perform, 3,x, sp[0]->s_vec, sp[0]->s_n);
+
+void rtpghi_tilde_recreatestate(t_rtpghi_tilde *x, LTFAT_FIRWIN win, ltfat_int blocksize, int overlap) {
+	if(x->sta_pd && (win != x->window || blocksize != x->blocksize || overlap != x->overlap)) {
+		rtpghi_done(&(x->sta_pd));
+		x->sta_pd=NULL;
+		post("sta zeroed");
+	}
+	if (!x->sta_pd) {
+		ltfat_int w = 1;
+		x->blocksize = blocksize;
+		x->window = win;
+		x->overlap = overlap;
+		double tol = x->tol_pd;
+		int do_causal = x->do_causal_pd;
+		double gamma;
+		
+		gamma = phaseret_firwin2gamma(win, x->blocksize);
+		if (rtpghi_init(w, blocksize/(x->overlap), x->blocksize, gamma, tol, do_causal, &(x->sta_pd))) {
+			pd_error(x, "failed to init state");
+			x->sta_pd=NULL;
+			}
+		}
 }
 
-void *rtpghi_tilde_new(t_symbol *s, int argc, t_atom *argv)
-{
+void rtpghi_tilde_dsp(t_rtpghi_tilde *x, t_signal **sp) {
+	//post("start DSP");
+	ltfat_int M = (ltfat_int) sp[0]->s_n;
+	if ((x->blocksize) == 0) {
+		(x->blocksize) = M;
+	}
+
+	if (x->window_type_pd != NULL) {
+		x->window = ltfat_str2firwin(x->window_type_pd->s_name);
+	}
```

Comment:
hmm, no. this doesn't really work as *i* imagined it.
if you set the members (`x->window`, `x->blocksize`) before calling `rtpghi_tilde_recreatestate()`,
the latter will not reallocate properly (as it now checks whether `x->window != x->window` which is obviously always true)